sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala

```diff
@@ -40,23 +40,47 @@ class EquivalentExpressions {
     override def hashCode: Int = e.semanticHash()
   }
 
+  /**
+   * Wrapper around an Expression that provides structural semantic equality.
+   */
+  case class StructuralExpr(e: Expression) {
+    def normalized(expr: Expression): Expression = {
+      expr.transformUp {
+        case b: ParameterizedBoundReference =>
```

Comment:
To avoid unnecessary plan copys, we can check this equality based on `BoundReference` (by just copying it like `b.copy(ordinal = 0 or -1?)`)? IIUC its ok to replace `BoundReference` with `ParameterizedBoundReference` just when generating code in https://github.com/apache/spark/pull/25717/files#diff-8bcc5aea39c73d4bf38aef6f6951d42cR1117?

Reply:
Isn't b.copy also copying the input expr?

Reply:
Ah, I see. But, if we write it like this (https://github.com/apache/spark/commit/c1c5052902641d0cabc0a74f1556f01d3b2f4234), we don't need to pass `CodegenContext` into `EquivalentExpressions`?

Reply:
Ok. looks good.

Reply:
I just modified like this way, but it failed HashAggregationQuerySuite in jenkins and local.

Reply:
I think I figured why. Made another comment.

Reply:
The current one looks super good to me. Thanks!